lib/src/main/kotlin/com/swmansion/starknet/data/TypedData.kt

```diff
@@ -80,48 +85,135 @@ import kotlinx.serialization.json.*
 @Suppress("DataClassPrivateConstructor")
 @Serializable
 data class TypedData private constructor(
-    val types: Map<String, List<TypeBase>>,
+    @SerialName("types")
+    val customTypes: Map<String, List<Type>>,
+
     val primaryType: String,
-    val domain: JsonObject,
+
+    val domain: Domain,
+
     val message: JsonObject,
 ) {
-    init {
-        val reservedTypeNames = listOf("felt", "felt*", "string", "string*", "selector", "selector*", "merkletree", "merkletree*", "raw", "raw*")
-        reservedTypeNames.forEach {
-            require(!types.containsKey(it)) { "Types must not contain $it." }
-        }
-    }
-
     constructor(
-        types: Map<String, List<TypeBase>>,
+        customTypes: Map<String, List<Type>>,
         primaryType: String,
         domain: String,
         message: String,
     ) : this(
-        types = types,
+        customTypes = customTypes,
         primaryType = primaryType,
-        domain = Json.parseToJsonElement(domain).jsonObject,
+        domain = Json.decodeFromString(domain),
         message = Json.parseToJsonElement(message).jsonObject,
     )
 
+    private val revision = domain.revision ?: Revision.V0
+
+    @Transient
+    val types: Map<String, List<Type>> = run {
+        val presetTypes = when (revision) {
+            Revision.V0 -> presetTypesV0
+            Revision.V1 -> presetTypesV1
+        }
+        customTypes + presetTypes
+    }
+
+    private val hashMethod by lazy {
+        when (revision) {
+            Revision.V0 -> HashMethod.PEDERSEN
+            Revision.V1 -> HashMethod.POSEIDON
+        }
+    }
+
+    init {
+        verifyTypes()
+    }
+
+    private fun hashArray(values: List<Felt>) = hashMethod.hash(values)
+
+    private fun verifyTypes() {
+        require(domain.separatorName in customTypes) { "Types must contain '${domain.separatorName}'." }
+
+        getBasicTypes(revision).forEach { require(it !in customTypes) { "Types must not contain basic types. [$it] was found." } }
+        getPresetTypes(revision).keys.forEach { require(it !in customTypes) { "Types must not contain preset types. [$it] was found." } }
+
+        val referencedTypes = customTypes.values.flatten().flatMap {
+            when (it) {
+                is EnumType -> extractEnumTypes(it.type) + it.contains
+                is MerkleTreeType -> listOf(it.contains)
+                is StandardType -> listOf(stripPointer(it.type))
+            }
+        }.distinct() + domain.separatorName + primaryType
+
+        customTypes.keys.forEach {
+            require(it.isNotEmpty()) { "Types cannot be empty." }
+            require(!it.isArray()) { "Types cannot end in *. [$it] was found." }
+            require(!it.startsWith("(") && !it.endsWith(")")) { "Types cannot be enclosed in parentheses. [$it] was found." }
```

Comment:
https://github.com/starknet-io/SNIPs/pull/77/files

Not sure if this should be interpreted as "cannot be enclosed in ()" or should also apply to a singular parenthesis

Reply:
I understand it as inside `()`